Sources/Tests/FeatureTests/pfPasswordStoreTest/test_pfPasswordStore.cpp

```diff
@@ -0,0 +1,127 @@
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include <gtest/gtest.h>
+#include <string_theory/string>
+#include "plFileSystem.h"
+#include "pnUUID/pnUUID.h"
+#include "pfPasswordStore/pfPasswordStore.h"
+#include "pfPasswordStore/pfPasswordStore_impl.h"
+
+TEST(pfPasswordStore, service)
+{
+    pfPasswordStore* store = pfPasswordStore::Instance();
+    EXPECT_NE(store, nullptr);
+
+#if defined(HS_BUILD_FOR_WIN32)
+    EXPECT_EQ(typeid(*store), typeid(pfWin32PasswordStore));
+#elif defined(HS_BUILD_FOR_APPLE)
+    EXPECT_EQ(typeid(*store), typeid(pfApplePasswordStore));
+#elif defined(HAVE_LIBSECRET)
+    EXPECT_EQ(typeid(*store), typeid(pfUnixPasswordStore));
+#else
+    EXPECT_EQ(typeid(*store), typeid(pfFilePasswordStore));
+#endif
+}
+
+#if defined(HS_BUILD_FOR_WIN32)
+TEST(pfWin32PasswordStore, storing_and_retrieval)
+{
+    pfWin32PasswordStore store;
+    ST::string username = "test_pfPasswordStore";
+    ST::string password = plUUID::Generate().AsString();
+
+    bool success = store.SetPassword(username, password);
+    EXPECT_EQ(success, true);
+
+    ST::string pass = store.GetPassword(username);
+    EXPECT_EQ(pass.compare(password), 0);
+}
+#endif
+
+#if defined(HS_BUILD_FOR_APPLE)
+TEST(pfApplePasswordStore, storing_and_retrieval)
+{
+    pfApplePasswordStore store;
+    ST::string username = "test_pfPasswordStore";
+    ST::string password = plUUID::Generate().AsString();
+
+    bool success = store.SetPassword(username, password);
+    EXPECT_EQ(success, true);
+
+    ST::string pass = store.GetPassword(username);
+    EXPECT_EQ(pass.compare(password), 0);
+}
+#endif
+
+#if defined(HAVE_LIBSECRET)
+TEST(pfUnixPasswordStore, storing_and_retrieval)
+{
+    pfUnixPasswordStore store;
+    ST::string username = "test_pfPasswordStore";
+    ST::string password = plUUID::Generate().AsString();
+
+    bool success = store.SetPassword(username, password);
+    EXPECT_EQ(success, true);
+
+    ST::string pass = store.GetPassword(username);
+    EXPECT_EQ(pass.compare(password), 0);
+}
+#endif
+
+TEST(pfFilePasswordStore, fallback_storing_and_retrieval)
+{
+    // Ensure we don't overwrite the real login.dat file
+    plFileSystem::CreateDir("init");
+    FILE* f = plFileSystem::Open(plFileName::Join("init", "login.dat"), "wb");
+    fclose(f);
```

Comment:
I'm not really sure I understand what's going on here, even with the comment.

Reply:
See https://github.com/H-uru/Plasma/blob/913276f4cd5bce8c5dbc78bed2d1f6ec6804d960/Sources/Plasma/FeatureLib/pfPasswordStore/pfPasswordStore.cpp#L117-L124

In External Release mode, it will read/write to the init/login.dat file in the user app data folder. In internal builds, it will use a local init/login.dat file if it exists.

For this test, I thought it would be best to ensure that local file exists, so there's a lower chance of people overwriting their login creds in user app data if they run the tests. It probably makes no difference for CI.

Reply:
I think the comment should be expanded to include that information... and maybe this test disabled on external releases? I'm unsure about that last bit. Or maybe

Reply:
I think the comment should be expanded to include that information. Or, perhaps the file password store should be modified such that we can override the hard coded login.dat file name in the case of tests.